src/blockstore/database.zig

```diff
@@ -0,0 +1,202 @@
+        ///    case, getBytes is faster than get. But if you *do* need an owned slice,
+        ///    then it's faster to call `get` insted of calling this function followed
+        ///    by memcpy.
+        pub fn getBytes(
+            self: *Self,
+            comptime cf: ColumnFamily,
+            key: cf.Key,
+        ) anyerror!?BytesRef {
+            return try self.impl.getBytes(cf, key);
+        }
+
+        pub fn delete(self: *Self, comptime cf: ColumnFamily, key: cf.Key) anyerror!void {
+            return try self.impl.delete(cf, key);
+        }
+
+        pub fn writeBatch(self: *Self) anyerror!WriteBatch {
+            return .{ .impl = self.impl.initBatch() };
+        }
+
+        pub fn commit(self: *Self, batch: WriteBatch) anyerror!void {
+            return self.impl.commit(batch.impl);
+        }
+
+        pub const WriteBatch = struct {
+            impl: Impl.WriteBatch,
+
+            pub fn put(
+                self: *WriteBatch,
+                comptime cf: ColumnFamily,
+                key: cf.Key,
+                value: cf.Value,
+            ) anyerror!void {
+                return try self.impl.put(cf, key, value);
+            }
+
+            pub fn delete(self: *WriteBatch, comptime cf: ColumnFamily, key: cf.Key) anyerror!void {
+                return try self.impl.delete(cf, key);
+            }
+        };
+    };
+}
+
+pub const ColumnFamily = struct {
+    name: []const u8,
+    Key: type,
+    Value: type,
+
+    const Self = @This();
+
+    /// At comptime, find this family in a slice. Useful for for fast runtime
+    /// accesses of data in other slices that are one-to-one with this slice.
+    pub fn find(comptime self: Self, comptime column_families: []const Self) comptime_int {
+        for (column_families, 0..) |column_family, i| {
+            if (std.mem.eql(u8, column_family.name, self.name)) {
+                return i;
+            }
+        }
+        @compileError("not found");
+    }
+};
+
+/// Bincode-based serializer that should be usable by database implementations.
+pub const serializer = struct {
+    /// Returned slice is owned by the caller. Free with `allocator.free`.
+    pub fn serializeAlloc(allocator: Allocator, item: anytype) ![]const u8 {
+        const buf = try allocator.alloc(u8, try sig.bincode.sizeOf(item, .{}));
+        return sig.bincode.writeToSlice(item, buf);
+    }
+
+    /// Returned data may or may not be owned by the caller.
+    /// Do both:
+    ///  - Assume the data is owned by the scope where `item` originated,
+    ///    so finish using the slice before returning from the caller (do not store slice as-is)
+    ///  - Call BytesRef.deinit before returning from the caller (as if you own it).
+    ///
+    /// Use this if the database backend accepts a pointer and immediately calls memcpy.
+    pub fn serializeToRef(allocator: Allocator, item: anytype) !BytesRef {
```

Comment:
this can probably be a bincode helper function - could see it being useful elsewhere too